clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -1103,8 +1103,37 @@ CIRGenModule::getConstantArrayFromStringLiteral(const StringLiteral *E) {
     return builder.getString(Str, eltTy, finalSize);
   }
 
-  assert(0 && "not implemented");
-  return {};
+  auto arrayTy =
+      getTypes().ConvertType(E->getType()).dyn_cast<mlir::cir::ArrayType>();
+  assert(arrayTy && "string literals must be emitted as an array type");
+
+  auto arrayEltTy = arrayTy.getEltType().dyn_cast<mlir::cir::IntType>();
+  assert(arrayEltTy &&
+         "string literal elements must be emitted as integral type");
+
+  auto arraySize = arrayTy.getSize();
+  auto literalSize = E->getLength();
+
+  // Collect the code units.
+  SmallVector<uint32_t, 32> elementValues;
```

Comment:
Does this cover both `ElemTy->getPrimitiveSizeInBits() == 16` and `ElemTy->getPrimitiveSizeInBits() == 32`? Seems like it does but you're approach is slightly different from traditional codegen, so I wonder.

(If not, this PR should add an assert for anything not fully supported.)

Reply:
Yes I believe they're all covered.

The original clang CodeGen just copies the same code to handle these two cases separately. I merges them together.